Client/core/Graphics/CRenderItem.EffectCloner.h

```diff
@@ -7,6 +7,37 @@
  *
  *****************************************************************************/
 
+struct SEffectInvariant
+{
+    SString      strFile;
+    EffectMacroList macros;
+
+    bool operator ==(const SEffectInvariant& e) const
+    {
+        return strFile == e.strFile && macros == e.macros;
+    }
+};
+
+template<>
+struct std::hash<SEffectInvariant>
+{
+    size_t operator()(const SEffectInvariant& entry) const
+    {
+        using EffectHash = std::size_t;
+
+        EffectHash uiHash = std::hash<SString>{}(entry.strFile);
+
+        for (const auto& [name, definition] : entry.macros)
+        {
+            // Hash function from boost
+            uiHash ^= std::hash<SString>{}(name)       + 0x9e3779b9 + (uiHash << 6) + (uiHash >> 2);
+            uiHash ^= std::hash<SString>{}(definition) + 0x9e3779b9 + (uiHash << 6) + (uiHash >> 2);
+        }
+
+        return uiHash;
+    }
+};
+
 ////////////////////////////////////////////////////////////////
```

Comment:
The old code was a lot simpler. Explicit template specialization is very helpful when you want different code to execute for different data types for the same struct or function. In this case, `CalculateInvariantHash` was only referenced in CEffectCloner::CreateD3DEffect. I don't see any benefit of using a template for something simple like this.  

In my opinion, it would be nice to bring back `CEffectCloner::CalculateInvariantHash` and  set the type of **m_ValidMap** to 
```
std::unordered_map<std::size_t, CEffectTemplate*> m_ValidMap;
```
The rest of the PR is good.

Reply:
It was a lot simpler, I agree. But without explicit element comparison(what unordered_map actually do) there is a risk of hash collision. This is why I added such a specialization. Complicated but reliable.